File: Server/DTO/Todo.cs
using System.ComponentModel.DataAnnotations;

namespace Server.DTO;

/// <summary>
/// TodoのDTOクラス
/// </summary>
public class Todo
{
    /// <summary>ID</summary>
    public int Id { get; set; }

    /// <summary>内容</summary>
    [Required]
    [MaxLength(100)]
    public string Content { get; set; } = string.Empty;

    /// <summary>予定日</summary>
    [DateObject.Validate]
    public DateObject DueDate { get; set; } = new DateObject(DateTime.Today);

    /// <summary>作成日</summary>
    [DateObject.Validate]
    public DateObject CreatedDate { get; set; } = new DateObject(DateTime.Now);

    /// <summary>完了日</summary>
    [DateObject.Validate(true)]
    public DateObject? CompletionDate { get; set; } = null;

    /// <summary>
    /// TodoモデルからTodoのDTOに変換します
    /// </summary>
    /// <param name="todo">Todoモデル</param>
    /// <returns>TodoのDTO</returns>
    public static DTO.Todo From(Models.Todo todo)
    {
        return new DTO.Todo
        {
            Id = todo.Id,
            Content = todo.Content,
            DueDate = new DateObject(todo.DueDate),
            CreatedDate = new DateObject(todo.CreatedDate),
            CompletionDate = todo.CompletionDate is DateTime completionDate ? new DateObject(completionDate) : null,
        };
    }
}

File: Client/src/Models/Todo.ts
import * as DTO from '../DTO'

/**
 * Todoのモデル
 */
export default class Todo {
    /** ID */
    id: number
    /** やること */
    content: string
    /** 予定日 */
    dueDate: Date
    /** 作成日 */
    createdDate: Date
    /** 完了日 */
    completionDate: Date | null

    /** コンストラクタ */
    constructor (id: number, content: string, dueDate: Date, createdDate: Date, completionDate: Date | null) {
        this.id = id
        this.content = content
        this.createdDate = createdDate
        this.dueDate = dueDate
        this.completionDate = completionDate
    }

    /**
     * サーバーから受信したTodoのDTOをTodoモデルに変換します
     * @param dto TodoのDTO
     * @param utcToLocale 日時データをutcからLocaleに変換するか
     * @returns Todoモデル
     */
    static from (dto: DTO.ITodo, utcToLocale: boolean = true) {
        return utcToLocale
            ? new Todo(
                dto.id,
                dto.content,
                dto.dueDate.toLocaleDate(),
                dto.createdDate.toLocaleDate(),
                dto.completionDate != null
                    ? dto.completionDate.toLocaleDate()
                    : dto.completionDate)
            : new Todo(dto.id, dto.content, dto.dueDate.toUTCDate(), dto.createdDate.toUTCDate(), dto.completionDate?.toUTCDate() ?? null)
    }
}
